ci-operator/step-registry/hypershift/kubevirt/create/hypershift-kubevirt-create-chain.yaml

```diff
@@ -0,0 +1,136 @@
+        --name ${KUBEVIRT_CLUSTER_NAME} \
+        --node-pool-replicas ${HYPERSHIFT_NODE_COUNT} \
+        --base-domain ${HYPERSHIFT_BASE_DOMAIN} \
+        --pull-secret=/etc/ci-pull-credentials/.dockerconfigjson \
+        --release-image ${RELEASE_IMAGE_LATEST} \
+
+      echo "Waiting for cluster to become available"
+      oc wait --timeout=30m --for=condition=Available --namespace=clusters hostedcluster/${KUBEVIRT_CLUSTER_NAME}
+      echo "Cluster became available, creating kubeconfig"
+      bin/hypershift create kubeconfig --namespace=clusters --name=${KUBEVIRT_CLUSTER_NAME} >${SHARED_DIR}/nested_kubeconfig
+
+      # Enable wildcard routes on the management cluster
+      oc patch ingresscontroller -n openshift-ingress-operator default --type=json -p \
+        '[{ "op": "add", "path": "/spec/routeAdmission", "value": {wildcardPolicy: "WildcardsAllowed"}}]'
+
+      export KUBECONFIG=${SHARED_DIR}/nested_kubeconfig
+
+      # Waiting for the default NodePort service to be created on the hosted cluster
+      until \
+        oc get service router-nodeport-default -n openshift-ingress
+      do
+        echo "router-nodeport-default service wasn't created yet on the hosted cluster."
+        sleep 10s
+      done
+
+      HTTPS_NODEPORT=$(oc get services -n openshift-ingress router-nodeport-default -o wide | awk '{print $5}' | awk -F "443:" '{print $2}' | awk -F "/" '{print $1}' | tr -d '[:space:]')
+      NESTED_BASE_DOMAIN=$(oc get ingresscontroller -n openshift-ingress-operator default -o yaml | grep domain | awk '{print $2 }')
+      CLUSTER_NAMESPACE=clusters-${KUBEVIRT_CLUSTER_NAME}
+      CLUSTER_BASE_DOMAIN=${KUBEVIRT_CLUSTER_NAME}.${NESTED_BASE_DOMAIN}
+
+      # Create a service for ingress to the nested cluster on the node's VMIs
+      oc create -f - <<EOF
+      apiVersion: v1
+      kind: Service
+      metadata:
+        labels:
+          app: ${KUBEVIRT_CLUSTER_NAME}
+        name: apps-ingress
+        namespace: ${CLUSTER_NAMESPACE}
+      spec:
+        internalTrafficPolicy: Cluster
+        ipFamilies:
+          - IPv4
+        ipFamilyPolicy: SingleStack
+        ports:
+          - name: https-443
+            port: 443
+            protocol: TCP
+            targetPort: ${HTTPS_NODEPORT}
+        selector:
+          kubevirt.io: virt-launcher
+        sessionAffinity: None
+        type: ClusterIP
+      EOF
+
+      # Create a Route for the ingress subdomain of the nested cluster
+      oc create -f - <<EOF
+      apiVersion: route.openshift.io/v1
+      kind: Route
+      metadata:
+        name: ${KUBEVIRT_CLUSTER_NAME}-443
+        namespace: ${CLUSTER_NAMESPACE}
+      spec:
+        host: data.apps.${CLUSTER_BASE_DOMAIN}
+        wildcardPolicy: Subdomain
+        tls:
+          termination: passthrough
+        port:
+          targetPort: https-443
+        to:
+          kind: Service
+          name: apps-ingress
+          weight: 100
+      EOF
+
```

Comment:
When i was debugging kubevirt guest clusters with the presubmit, i found that having logic that waits for a few intermediate steps (before checking that all operators came online) was helpful

This included things like..

- all vmis in CLUSTER_NAMESPACE to reach a `vmi.Status.Phase=Running`
- all expected nodes in the guest cluster to be detected, so nodeCount == replica count. This just verifies that the vmis came online and could contact the control plane at startup.

That gave me a clue if something was breaking down in the process before the operators had a chance to come online within the guest cluster.

Reply:
I added these two additional checks, as you suggested.

1. `oc wait -n ${CLUSTER_NAMESPACE} vmi --all --for=jsonpath='{.status.phase}'=Running --timeout 30m` - to verify all of the VMIs are running.
2. comparing `.spec.replicas` of the nodepool of the nested cluster, to the actual number of nodes in the nested cluster itself.

This should be robust enough to ensure the kubevirt infra of the nested cluster is ok before waiting for its ClusterOperators to be all available.